llvm/lib/Target/MSP430/MSP430FrameLowering.cpp

```diff
@@ -200,6 +200,31 @@ bool MSP430FrameLowering::spillCalleeSavedRegisters(
   return true;
 }
 
+const char *MSP430FrameLowering::tailCallNameForRestoringCalleSavedRegisters(MutableArrayRef<CalleeSavedInfo> CSI) const {
+  static unsigned Registers[] = {
+    /*MSP430::R4, ??? */ MSP430::R5, MSP430::R6, MSP430::R7,
+    MSP430::R8, MSP430::R9, MSP430::R10,
+  };
+  static const char *RestorerNames[] = {
```

Comment:
Can you provide the assembler code in the comments to illustrate the set of labels?

Reply:
Do you mean prepending this array with comment lines containing plain assembler code or "inlining" these comments right after label names?

Reply:
Prepending (or, rather, appending) looks ok. Something like:
```
"__mspabi_func_epilog_7", // __mspabi_func_epilog_7: POP R4
"__mspabi_func_epilog_6", //__mspabi_func_epilog_6: POP R5
```

Reply:
Another thing to consider: after that change, I would have two arrays: `Registers` and `RestorerNames` that seem to be logically aligned with each other, so it seems reasonable to convert them to a single array of `struct`s or so. In other projects I have seen using this together with parameterized `#define`s. Here it seems quite handy to define something like

```cpp
struct RestorerNameMapping {
  const char *name;
  unsigned reg;
};
static RestorerMapping Restorers[] = {
#define DEF(name_suff, reg) { \
    .name = "__mspabi_func_epilog_" name_suff, \
    .reg = reg \
  }
DEF("7", MSP430::R4),
DEF("6", MSP430::R5),
DEF("5", MSP430::R6),
DEF("4", MSP430::R7),
DEF("3", MSP430::R8),
DEF("2", MSP430::R9),
DEF("1", MSP430::R10),
#undef DEF
};
```

On one hand, this can make it harder to read. On the other hand, this eliminates some pitfalls:
* no way to mis-spell _one of_ `__mspabi_func_epilog` prefixes while retaining these strings in the form of literal constants (so no heap allocation or possibly dangling pointers to stack-allocated buffers)
* the two arrays have the same size by construction (because now these are a single array)
* after the `DEF` macro is ultimately deciphered by the reader, it should be much simpler to review this code due to less boilerplate

Reply:
Oops, this is C++ after all, so no `#define`s even required...